Expand paths with tilde

variadico opened the pull request:
Currently, if you pass a path with a tilde to nats.UserCredentials it fails
because it doesn't expand the tilde to mean the user's home directory.

This change adds support to expand paths with a tilde, such as "~/.nkeys".

Resolves https://github.com/nats-io/nats.go/issues/509

variadico commented:
No, the logic isn't complex. However, the dep seems to handle a lot of edge cases.
https://github.com/mitchellh/go-homedir/blob/master/homedir.go

If we only want to solve the simple case, then we could probably just do this.
```
if strings.HasPrefix(path, "~") {
    path = filepath.Join(os.Getenv("HOME"), path[1:])
}
```

Maybe at some point we'll get an issue saying that there is a bug in the expansion logic. Or maybe not.

Another option is that we could just copy the file over into the nats repo (with all the license stuff).

wallyqs commented:
IMO checking whether $HOME or $home are being defined in env when `~` is the first char from the path would good enough, then can return an error saying explicitly saying that $HOME has to be defined for it to work.

ripienaar commented:
In my experience this handles the bulk of cases, hashi one is undeniably better but this is good enough for me. 


```
func homeDir() (string, error) {
	if runtime.GOOS == "windows" {
		drive := os.Getenv("HOMEDRIVE")
		home := os.Getenv("HOMEDIR")

		if home == "" || drive == "" {
			return "", fmt.Errorf("Cannot determine home dir, ensure HOMEDRIVE and HOMEDIR is set")
		}

		return filepath.Join(os.Getenv("HOMEDRIVE"), os.Getenv("HOMEDIR")), nil
	}

	home := os.Getenv("HOME")

	if home == "" {
		return "", fmt.Errorf("Cannot determine home dir, ensure HOME is set")
	}

	return home, nil
}
```

derekcollison commented:
I prefer to limit the dependencies unless we absolutely have to have them.

variadico commented:
All right. Implemented `homeDir`. Added tests.
https://github.com/nats-io/nats.go/pull/512/commits/80090aff5e243283fc924b23fdace3d3fd57c0eb

Wasn't sure if I should add tests for windows as well... I probably should, right?

variadico commented:
Oookay, so turns out this homedir function is already in the stdlib.
https://golang.org/pkg/os/#UserHomeDir

The other one in `os/user` is bad because it uses cgo. This one in `os`, basically does what we were doing, i.e. just checking env vars.

So, I deleted the original function we had. I also cross compiled from Linux to Windows to verify `os.UserHomeDir` worked. It correctly printed the home dir in command prompt, powershell, and the linux subsystem.

One weird thing is I'm not sure how to expect Windows users to give us a path. I added some Windows tests too, hopefully @ColinSullivan1 can take a look at those to see if they make sense.

variadico commented on nats_test.go:
```diff
@@ -98,6 +98,94 @@ func TestVersionMatchesTag(t *testing.T) {
+
+			// missing USERPROFILE env var
+			{path: "~/fizz", wantErr: true},
+		}
```
Hey, @ColinSullivan1. Are these paths normally how Windows folks would give us a path?

ColinSullivan1 replied:
Not normally, unless running in some type of non-standard shell.  paths would be \\foo\\bar, wtih backslashes (escaped for the test).

derekcollison commented on nats.go:
```diff
@@ -4016,14 +4018,32 @@ func wipeSlice(buf []byte) {
+	}
+
+	home, err := os.UserHomeDir()
+	if err != nil {
```
Should this just have ~/ -> /?

variadico replied:
Hm? Sorry, don't follow.

derekcollison replied:
if the homedir return an err should we just remove the ~?

derekcollison commented on nats_test.go:
```diff
@@ -98,6 +98,94 @@ func TestVersionMatchesTag(t *testing.T) {
+func TestExpandTildePath(t *testing.T) {
+	t.Run("unix", func(t *testing.T) {
+		if runtime.GOOS == "windows" {
+			t.Skip()
```
Let's see if we can get one of these from @ColinSullivan1

ColinSullivan1 replied:
Will do...

ColinSullivan1 commented on nats.go:
```diff
@@ -4016,14 +4018,32 @@ func wipeSlice(buf []byte) {
+		return p, nil
+	}
+
+	home, err := os.UserHomeDir()
```
Golang doesn't check for `%HOMEDRIVE%%HOMEPATH%` which is also used.  It'd be more complicated but great for users if env var expansion in general was supported.

derekcollison commented:
Congrats! First PR merge!
